src/main/host/syscall/handler/mod.rs

```diff
@@ -268,48 +268,41 @@ impl SyscallHandler {
-            && !matches!(rv, Err(SyscallError::Blocked(_)))
-        {
-            let max_unapplied_cpu_latency = ctx.host.shim_shmem().max_unapplied_cpu_latency;
+        if ctx.process.is_running() && !matches!(rv, Err(SyscallError::Blocked(_))) {
+            let host_shmem = ctx.host.shim_shmem();
+            let mut host_shmem_prot = ctx.host.shim_shmem_lock_borrow_mut().unwrap();
 
             // increment unblocked syscall latency, but only for non-shadow-syscalls, since the
             // latter are part of Shadow's internal plumbing; they shouldn't necessarily "consume"
             // time
-            if !is_shadow_syscall(syscall) {
-                ctx.host
-                    .shim_shmem_lock_borrow_mut()
-                    .unwrap()
-                    .unapplied_cpu_latency += ctx.host.shim_shmem().unblocked_syscall_latency;
+            if ctx.host.shim_shmem().model_unblocked_syscall_latency && !is_shadow_syscall(syscall)
+            {
+                host_shmem_prot.unapplied_cpu_latency += host_shmem.unblocked_syscall_latency;
             }
 
-            let unapplied_cpu_latency = ctx
-                .host
-                .shim_shmem_lock_borrow()
-                .unwrap()
-                .unapplied_cpu_latency;
-
             log::trace!(
                 "Unapplied CPU latency amt={}ns max={}ns",
-                unapplied_cpu_latency.as_nanos(),
-                max_unapplied_cpu_latency.as_nanos()
+                host_shmem_prot.unapplied_cpu_latency.as_nanos(),
+                host_shmem.max_unapplied_cpu_latency.as_nanos()
             );
 
-            if unapplied_cpu_latency > max_unapplied_cpu_latency {
-                let new_time = Worker::current_time().unwrap() + unapplied_cpu_latency;
+            if host_shmem_prot.unapplied_cpu_latency > host_shmem.max_unapplied_cpu_latency {
+                let new_time = Worker::current_time().unwrap()
+                    + core::mem::replace(
+                        &mut host_shmem_prot.unapplied_cpu_latency,
+                        SimulationTime::ZERO,
+                    );
                 let max_time = Worker::max_event_runahead_time(ctx.host);
 
                 if new_time <= max_time {
-                    log::trace!("Reached unblocked syscall limit; Incrementing time");
-
-                    ctx.host
-                        .shim_shmem_lock_borrow_mut()
-                        .unwrap()
-                        .unapplied_cpu_latency = SimulationTime::ZERO;
+                    log::trace!(
+                        "Reached max-unapplied-cpu-latency, but not max runahead; Incrementing time"
+                    );
                     Worker::set_current_time(new_time);
                 } else {
-                    log::trace!("Reached unblocked syscall limit; Yielding");
+                    log::trace!(
+                        "Reached max-unapplied-cpu-latency, and max runahead; Rescheduling"
+                    );
```

Comment:
These two cases seem like they behave differently. If `new_time <= max_time`, then we update the current time for all threads of all processes on this host. But if `new_time > max_time`, we block the current thread and other threads and processes continue running at the current time. I wonder if we should set the current time to `max_time` here?

Maybe this is outside the scope of this MR. Is it worth opening an issue for?

Reply:
Added comments to try to clarify -

when `new_time < Worker::max_event_runahead_time`, we can safely bump the worker's time to `new_time`.

When it's not (because another scheduled event or the end of round occurs before `new_time`), we can't safely bump the worker's time to `new_time`, so we block and schedule ourselves to run at `new_time` instead.

i.e. 

> If new_time <= max_time, then we update the current time for all threads of all processes on this host.

Correct, but this condition only holds if no other threads or processes on the host would actually run before `new_time`.